src/plugins/discover/public/application/view_components/canvas/discover_visualization.tsx

```diff
@@ -47,36 +51,33 @@ export const DiscoverVisualization = ({ hits, bucketInterval, chartData, rows }:
     loadExpression();
   }, [toExpression, searchContext, rows, indexPattern, services]);
 
-  useLayoutEffect(() => {
+  useEffect(() => {
     const subscription = services.data.query.state$.subscribe(({ state }) => {
       setSearchContext({
         query: state.query,
         timeRange: state.time,
         filters: state.filters,
       });
+      setEnableViz(
+        queryString.getLanguageService().getLanguage(state.query!.language)!.showVisualization ??
+          false
+      );
     });
 
     return () => {
       subscription.unsubscribe();
     };
-  }, [services.data.query.state$]);
+  }, [queryString, services.data.query.state$]);
 
-  return (
+  return enableViz && expression ? (
     <EuiPanel className="discoverVisualization" data-test-subj="visualizationLoader">
-      {expression ? (
-        <ReactExpressionRenderer
-          key={JSON.stringify(searchContext) + expression}
-          expression={expression}
-          searchContext={searchContext}
-        />
-      ) : (
-        <EuiFlexItem
-          className="discoverVisualization__empty"
-          data-test-subj="emptyDiscoverVisualization"
-        >
-          <>No data to display</>
-        </EuiFlexItem>
-      )}
+      <ReactExpressionRenderer
+        key={JSON.stringify(searchContext) + expression}
+        expression={expression}
+        searchContext={searchContext}
```

Comment:
Just mirroring what I mentioned previously. It looks like we may want to create a new `searchContext` object when it gets updated so that React can do non-forced re-renders based on the shallow comparision.